### PR TITLE
PROD-364 Add UTM attributes in mixpanel

### PR DESCRIPTION
**Pull request by neel-ds**
- Description
Add UTM parameters in mixpanel track requests
- What are the steps to test that this code is working?
Add utm params in the url and verify the event in mixpanel. For e.g I added `utm_source` in an event
- Screen shots or recordings for UI changes
Screenshot 2024-10-14 at 4 49 15 PM

### CODE REVIEW

**Review by marvinmarnold — CHANGES_REQUESTED** [review]
The [docs](https://docs.mixpanel.com/docs/tracking-best-practices/server-side-best-practices#tracking-utms-and-referrer) list a few other UTM fields: `utm_content` and `utm_term`. Any reason you didn't include those?

**Comment by neel-ds** [comment]
> The [docs](https://docs.mixpanel.com/docs/tracking-best-practices/server-side-best-practices#tracking-utms-and-referrer) list a few other UTM fields: `utm_content` and `utm_term`. Any reason you didn't include those?

I have added those attributes now. ✅

**Review by marvinmarnold — DISMISSED** [review]

File: lib/mixpanel.ts [hunk 1]
```diff
@@ -10,6 +11,11 @@ const sendToMixpanel = async (
       : undefined,
     $screen_height: window.screen.height,
     $screen_width: window.screen.width,
+    $utm_source: urlParams.get("utm_source") || undefined,
+    $utm_medium: urlParams.get("utm_medium") || undefined,
+    $utm_campaign: urlParams.get("utm_campaign") || undefined,
+    $utm_term: urlParams.get("utm_term") || undefined,
+    $utm_content: urlParams.get("utm_content") || undefined,
```
[the remark this answers is not in the document]
Comment (Yihwan) [reply lib/mixpanel.ts:18]: actually scond thought, let's just keep to officially supported utms for now and add the arbitrary ones if needed.

**Review by Yihwan — CHANGES_REQUESTED** [review]
need to preserve utm attribution on route changes as noted here https://linear.app/gator/issue/PROD-364/add-utm-attribution-to-mixpanel#comment-29acb1fc

**Comment by neel-ds** [comment]
This PR adds initial_utm and last_utm tracking facilities in mixpanel route using Vercel KV. 

cc: @marvinmarnold @Yihwan 
Screenshot 2024-10-14 at 4 49 15 PM

File: app/api/mixpanel/route.ts [hunk 2]
```diff
@@ -39,22 +44,76 @@ export async function POST(request: NextRequest) {
...
+      let lastUtm = storedUtmData?.last_utm || {};
+
+      // Check if any new UTM parameters are provided
+      const hasNewUtmParams = Object.values(utmParams).some(
+        (value) => value !== undefined,
+      );
+
+      // Update initial and last UTM data
+      if (!storedUtmData || hasNewUtmParams) {
+        initialUtm = initialUtm.utm_source ? initialUtm : utmParams;
```
Comment (marvinmarnold) [inline app/api/mixpanel/route.ts:78]: How about: `Object.keys(initialUtm).length > 0 ? initialUtm : utmParams;`

File: app/api/mixpanel/route.ts [hunk 3]
```diff
@@ -39,22 +44,76 @@ export async function POST(request: NextRequest) {
...
+        lastUtm = { ...lastUtm, ...utmParams };
```
Comment (marvinmarnold) [inline app/api/mixpanel/route.ts:79]: `lastUtm = hasNewUtmParams ?  utmParams : lastUtm` because we don't want to mix UTM params from different events.

**Review by marvinmarnold — CHANGES_REQUESTED** [review]
There's also some important follow up: https://linear.app/gator/issue/PROD-364/add-utm-attribution-to-mixpanel

File: app/api/mixpanel/route.ts (hunk 2 above)
Comment (neel-ds) [reply app/api/mixpanel/route.ts:78]: Good catch, thank you!

File: app/api/mixpanel/route.ts (hunk 3 above)
Comment (neel-ds) [reply app/api/mixpanel/route.ts:79]: yes, make sense

**Review by marvinmarnold — CHANGES_REQUESTED** [review]
Please also add test cases.

**Review by Yihwan — APPROVED** [review]
thanks confirmed working. out of curiosity dyk why last_utm_source and Initial UTM Source are formatted differently in mixpanel? is it something that mixpanel does under the hood?

**Comment by neel-ds** [comment]
> thanks confirmed working. out of curiosity dyk why last_utm_source and Initial UTM Source are formatted differently in mixpanel? is it something that mixpanel does under the hood?

To know the origin of first lead and track last touch of a user. It was proposed by Marvin and more details can be found in Linear issue comments.

**Pull request merged**